### ISSUE TITLE
linter: ESLint plugins that don't default export fail to load

### ISSUE THREAD

**Issue by tido64** [oxc-project/oxc#19958]
While migrating from ESLint, we hit an issue where a plugin wouldn't load because oxlint assumes all plugins default export:

https://github.com/oxc-project/oxc/blob/c92422bb0f51e039c10b7afa0b26382ad3cbdf84/apps/oxlint/src-js/plugins/load.ts#L128

But as you can see here, `eslint-plugin-wdio` only has named exports: https://github.com/webdriverio/webdriverio/blob/f6ca37e8c10beda004a11d685e7407b58aa22cb6/packages/eslint-plugin-wdio/src/index.ts#L48-L52

Would it be appropriate to change the load function to do something like this instead:

```ts
    const m = await import(url);
    const plugin = (m.default ?? m) as Plugin;
```

**Comment by overlookmotel** [comment]
Sorry for slow reply. Quite a sprint to get the alpha release out.

> Would it be appropriate to change the load function to do something like this instead:
> 
>     const m = await import(url);
>     const plugin = (m.default ?? m) as Plugin;

I don't think this would work. The package doesn't actually export the plugin directly, it only exports configs.

I think to support Oxlint the package should either:

1. Export the plugin itself as default.
2. Add a subpath export for the plugin object so you can access it directly as `eslint-plugin-wdio/plugin`.

Perhaps you'd like to make a PR on their repo to propose one of those?

### Workaround

Create a plugin that wraps `eslint-plugin-wdio`:

```js
// wdio-plugin.mjs
import { configs } from "eslint-plugin-wdio";

const plugin = configs['flat/recommended'].plugins.wdio.
plugin.meta = { name: "eslint-plugin-wdio" };

export default plugin;
```

Then use that wrapper plugin in your config:

```json
{
  "jsPlugins": ["./path/to/wdio-plugin.mjs"],
  "rules": {
    "wdio/await-expect": "error",
  }
}
```

### Side note

I noticed that `eslint-plugin-wdio` automatically enables TS-ESLint's `no-floating-promises` rule if you have TS-ESLint package installed. That's won't work in Oxlint (we don't currently support type-awareness in JS plugins) and, even if it did work, [the native version of this rule built in to Oxlint](https://oxc.rs/docs/guide/usage/linter/type-aware.html) will be much faster.

**Comment by overlookmotel** [comment]
We could consider adding `export` and `property` properties to aliases to support plugin packages like this:

```json
{
  "jsPlugins": [
    {
      "name": "wdio",
      "specifier": "eslint-plugin-wdio",
      "export": "configs",
      "property": ["flat/recommended", "plugins", "wdio"]
    }
  ]
}
```

Loading logic would become:

```js
let plugin = await import(url);

plugin = config.export != null ? plugin[config.export] : plugin.default;

const { property } = config;
if (property != null) {
  if (typeof property === "string" || typeof property === "number") {
    plugin = plugin[property];
  } else if (Array.isArray(property)) {
    for (const part of property) {
      plugin = plugin[part];
    }
  } else {
    throw new Error("Invalid `property`");
  }
}
```

But that's a complication to our API, and not the nicest config for user either!

Before going ahead with that, I think we should wait and see if any other plugins have this same "doesn't actually export the plugin directly" problem. I imagine it's rare because it forces you to use one of the default configs as starting point - doesn't allow you to start with just the plugin and enable rules that you specify.

**Comment by tido64** [comment]
Thanks, but it looks like `eslint-plugin-wdio` fixed this in 9.26: https://github.com/webdriverio/webdriverio/commit/a66e28fba1b063e5e423e791ac784550b39b41ce

Please feel free to close this issue.

**Comment by overlookmotel** [comment]
Thanks for the update @tido64.
